docs/key-issues/gpus.md

```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+some driver version X is installed will always work correctly when run with
+some newer driver version Y>X. As briefly discussed above, as of CUDA 11.0
+CUDA also promises [minor version
+compatibility](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility)
+(MVC). This compatibility guarantees that CUDA code compiled using a certain
+version of the CUDA toolkit will be compatible with any driver version within
+the same major release. This behavior is useful because it is often easier for
+users to upgrade their CUDA runtime than it is to upgrade the driver,
+especially on shared machines. For instance, the CUDA toolkit may be installed
+using conda, while the driver library cannot be. An example of leveraging MVC
+would be compiling code against the CUDA 11.5 runtime library and then running
+it on a system with a CUDA 11.2 driver installed. However there are some
+caveats with MVC:
+
+- If the code uses any features that were introduced in a later driver version
+  than the installed version, it will still fail to run. However, it will be a
+  runtime failure in the form of a `cudaErrorCallRequiresNewerDriver` CUDA
+  error, rather than a linker error or some similarly opaque issue. One
+  solution to this problem is for libraries to use runtime checks of the CUDA
+  version (using e.g. `cudaDriverGetVersion`) to only use supported features on
+  the installed driver.
+- NVRTC did not start supporting MVC until CUDA 11.3. Therefore, code that uses
+  NVRTC for JIT compilation must have been compiled with a CUDA version >= 11.3.
+- MVC only applies to CUDA code, not
+  [PTX](https://docs.nvidia.com/cuda/parallel-thread-execution/). PTX is an
+  instruction set that the CUDA driver library can JIT-compile to machine
+  instructions. The standard [CUDA compilation
+  pipeline](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#the-cuda-compilation-trajectory)
+  includes the translation of CUDA code into PTX, but in addition developers
+  may write PTX code directly rather than CUDA for performance since it offers
+  avenues for optimization that are either unavailable or very difficult to
+  access with CUDA C(++). However, since PTX code does not support MVC, PTX
+  compiled with a particular CUDA runtime may not work if run on a system with
+  an older driver. This fact has two consequences. First, libraries that
+  package PTX code will not benefit from MVC. Second, libraries that leverage
+  any sort of JIT-compilation pipeline that generates PTX code will _also_ not
+  support MVC. The latter can lead to more surprising behaviors, such as if a
+  user has a newer CUDA runtime than driver and then uses
+  [numba.cuda](https://numba.pydata.org/) to compile a Python function since
+  numba compiles CUDA kernels to PTX as part of its pipeline. Prior to CUDA
+  12, CUDA itself provides no solutions to this problem, although in some cases
```

Comment:
I don't think this is strictly true - since 11.1 or 11.2 (I can't recall exactly when) `libnvptxcompiler_static.a` has been bundled with the toolkit in order to support JIT compilation of PTX for older drivers than the toolkit used for the build - the build links in `libnvptxcompiler_static.a` and uses it to compile PTX bundled or generated at runtime. [PTX Compiler APIs](https://docs.nvidia.com/cuda/ptx-compiler-api/index.html) documents the usage of the library and API. The addition RAPIDS made on top of this was to build a Pythonic binding capable of patching Numba to use its bundled version of libnvptxcompiler_static which we build with 11.8, so it can always compile any 11.x PTX (because 11.8 was the last 11.x release).

This is only a partial solution - if multiple PTX files need to be linked, then there isn't an official general solution, but RAPIDS also provides the cubinlinker package, which provides a Pythonic binding to some internal APIs to address both PTX compilation and linking.

Reply:
True, there are some partial solutions available. IMO the distinction between compiling a single PTX file and linking multiple is too deep in the weeds even for this document. I personally wouldn't even know how to describe exactly when the former is sufficient and when the latter would be necessary. I'd love to fill that gap in my own knowledge, but I think adding that here at this point would be a bit too overwhelming for readers who aren't familiar with the entire CUDA compilation pipeline.

Do you think there's a reasonable middle ground in what can be stated here? In particular, if we focus on it from the perspective of a developer trying to make their package run: for what use cases are the PTX compiler APIs sufficient, and when would something like cubinlinker be required?

Reply:
> Is it accurate to state that this basically means that PTX functions cannot call PTX functions from another TU?

It's accurate, but it also doesn't elaborate on the implications. That TU might be from another PTX file provided by the user, or it might be in a library distributed as part of the toolkit, e.g. `libcudadevrt.a`, which contains functions used e.g. for dynamic parallellism or cooperative groups. So it also means that certain CUDA functionality becomes unavailable.

Reply:
> Can CUDA source code from another TU call PTX functions compiled using the PTX compiler API, or is the limitation bidirectional?

The limitation is bidirectional, because you can't add anything compiled for MVC to a link.

Reply:
> I assume that the latter works, otherwise it's fairly useless to have a completely isolated compilation unit that is JIT-compiled but can't be used by anything else.

The majority of Numba functionality can be supported in this way, it's not completely useless :slightly_smiling_face: - just not cooperative groups and linking C/C++ device code.

> I'd love to fill that gap in my own knowledge, but I think adding that here at this point would be a bit too overwhelming for readers who aren't familiar with the entire CUDA compilation pipeline.

Indeed, it is very complicated to spell out exactly what is and isn't possible, especially in generic terms and not "which subset of Numba can work in this case".

> Do you think there's a reasonable middle ground in what can be stated here? In particular, if we focus on it from the perspective of a developer trying to make their package run: for what use cases are the PTX compiler APIs sufficient, and when would something like cubinlinker be required?

I think maybe the middle ground just looks like this:

> CUDA itself provides no **general** solutions to this problem

(addition of the word "general") and leaving it at that, otherwise we end up discussing all these nuances and losing the wood for trees.

Reply:
OK sounds good. Done.

Reply:
I'm confused at how numba works given the bidirectionality. If a numba-compiled kernel can't link to anything or be linked to, how does anything call it?